src/shared/mount-util.c

```diff
@@ -490,52 +490,81 @@ int mount_move_root(const char *path) {
         return RET_NERRNO(chdir("/"));
 }
 
-int mount_pivot_root(const char *path) {
+int mount_pivot_root(const char *path, MountAttrPropagationType type) {
         _cleanup_close_ int fd_oldroot = -EBADF, fd_newroot = -EBADF;
+        unsigned int flags;
+        const char *prop_name;
 
         assert(path);
 
-        /* pivot_root() isn't currently supported in the initramfs. */
-        if (in_initrd())
-                return mount_move_root(path);
-
-        fd_oldroot = open("/", O_PATH|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
-        if (fd_oldroot < 0)
-                return log_debug_errno(errno, "Failed to open old rootfs");
-
-        fd_newroot = open(path, O_PATH|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
-        if (fd_newroot < 0)
-                return log_debug_errno(errno, "Failed to open new rootfs '%s': %m", path);
-
-        /* Change into the new rootfs. */
-        if (fchdir(fd_newroot) < 0)
-                return log_debug_errno(errno, "Failed to change into new rootfs '%s': %m", path);
-
-        /* Let the kernel tuck the new root under the old one. */
-        if (pivot_root(".", ".") < 0)
-                return log_debug_errno(errno, "Failed to pivot root to new rootfs '%s': %m", path);
-
-
-        /* At this point the new root is tucked under the old root. If we want
-         * to unmount it we cannot be fchdir()ed into it. So escape back to the
-         * old root. */
-        if (fchdir(fd_oldroot) < 0)
-                return log_debug_errno(errno, "Failed to change back to old rootfs: %m");
-
-        /* Note, usually we should set mount propagation up here but we'll
-         * assume that the caller has already done that. */
+        switch (type) {
+        case PROPAGATION_INHERIT:
+                flags = 0;
+                prop_name = "inherited";
+                break;
+        case PROPAGATION_PRIVATE:
+                flags = MS_PRIVATE;
+                prop_name = "private";
+                break;
+        case PROPAGATION_DEPENDENT:
+                flags = MS_SLAVE;
+                prop_name = "dependent";
+                break;
+        case PROPAGATION_SHARED:
+                flags = MS_SHARED;
+                prop_name = "shared";
+                break;
+        default:
+                assert_not_reached();
+        }
 
-        /* Get rid of the old root and reveal our brand new root. */
-        if (umount2(".", MNT_DETACH) < 0)
-                return log_debug_errno(errno, "Failed to unmount old rootfs: %m");
+        if (in_initrd()) {
+                int r;
```

Comment:
Please declare the special variable `r` at the beginning of the function even if it is used at only this block.

Reply:
Is it a rule that we only do function scope? And if not what does this give us if we move it to the top?

Reply:
See https://systemd.io/CODING_STYLE/.
> Using C Constructs
>
> Allocate local variables where it makes sense: at the top of the block, or at the point where they can be initialized. Avoid huge variable declaration lists at the top of the function.
>
> As an exception, `r` is typically used for a local state variable, but should almost always be declared as the last variable at the top of the function.

Reply:
"almost always" it would only be used for a single statement and the sentence above mentions that they should be declared at the point were initialized. Both would suggest that it's appropriate were it was placed.

It becomes a mute point though since `r` isn't present anymore once I fold mount_move_root() into the other function.